src/core_plugins/kibana/public/dashboard/dashboard_state.js

```diff
@@ -10,39 +13,114 @@ import { getPersistedStateId } from 'plugins/kibana/dashboard/panel/panel_state'
-    this.stateDefaults = getStateDefaults(dashboard);
+  constructor(dashboard, AppState) {
+    this.dashboard = dashboard;
+
+    this.stateDefaults = getStateDefaults(this.dashboard);
 
     this.appState = new AppState(this.stateDefaults);
     this.uiState = this.appState.makeStateful('uiState');
-    this.timefilter = timefilter;
-    this.dashboard = dashboard;
+    this.isDirty = false;
 
-    this.stateMonitor = stateMonitorFactory.create(this.appState, this.stateDefaults);
-
-    if (this.getShouldSyncTimefilterWithDashboard() && timeFilterPreviouslyStored) {
-      this.syncTimefilterWithDashboard(quickTimeRanges);
-    }
+    // We can't compare the filters stored on this.appState to this.dashboard because in order to apply
+    // the filters to the visualizations, we need to save it on the dashboard. We keep track of the original
+    // filter state in order to let the user know if their filters changed and provide this specific information
+    //in the 'lose changes' warning message.
+    this.lastSavedDashboardFilters = this.getFilterState();
 
     PanelUtils.initPanelIndexes(this.getPanels());
+    this.createStateMonitor();
+  }
+
+  /**
+   * Resets the state back to the last saved version of the dashboard.
+   */
+  resetState() {
+    // appState.reset uses the internal defaults to reset the state, but some of the default settings (e.g. the panels
+    // array) point to the same object that is stored on appState and is getting modified.
+    // The right way to fix this might be to ensure the defaults object stored on state is a deep
+    // clone, but given how much code uses the state object, I determined that to be too risky of a change for
+    // now.  TODO: revisit this!
+    this.stateDefaults = getStateDefaults(this.dashboard);
+    // The original query won't be restored by the above because the query on this.dashboard is applied
+    // in place in order for it to affect the visualizations.
+    this.stateDefaults.query = this.lastSavedDashboardFilters.query;
+    // Need to make a copy to ensure they are not overwritten.
+    this.stateDefaults.filters = Object.assign(new Array(), this.getLastSavedFilterBars());
+
+    this.isDirty = false;
+    this.appState.setDefaults(this.stateDefaults);
+    this.appState.reset();
+    this.stateMonitor.setInitialState(this.appState.toJSON());
+  }
+
+  /**
+   * Returns an object which contains the current filter state of this.dashboard.
+   * @returns {{timeTo: String, timeFrom: String, filterBars: Array, query: Object}}
+   */
+  getFilterState() {
+    return {
+      timeTo: this.dashboard.timeTo,
+      timeFrom: this.dashboard.timeFrom,
+      filterBars: this.getDashboardFilterBars(),
+      query: this.getDashboardQuery()
+    };
+  }
+
+  getTitle() {
+    return this.appState.title;
+  }
+
+  setTitle(title) {
+    this.appState.title = title;
+    this.saveState();
   }
 
   getAppState() {
```

Comment:
Agreed, it was the only thing I could come up with to prevent having to expose the AppState itself, but it's not the most readable, and the whole `apply` naming could make it rather confusing in the context of Javascript because of `Function.prototype.apply`